File: 2º Periodo/AEDS-II/Códigos/C/listaFlexivel.c
#include <stdio.h>
#include <stdlib.h>

//classe da celula
typedef struct Celula{
    
    int elemento;
    struct Celula* prox;

}Celula;

Celula* NovaCelula(int elemento){
    
    Celula* nova = (Celula*) malloc(sizeof(Celula));
    nova -> elemento = elemento; // em java this.elemento = elemento;
    nova -> prox = NULL; // em java prox = null;
    return nova;   
}

//criando os pornteiros da lista
Celula* primeiro;
Celula* ultimo;



void start(){
    
   primeiro = NovaCelula(-1);
   ultimo = primeiro;
 
}


void inserirFim(int x){
    
    ultimo->prox = NovaCelula(x);
    ultimo = ultimo->prox;
    
}


void inserirInicio(int x){
    
    Celula* tmp = NovaCelula(x);
    tmp->prox = primeiro->prox;
    primeiro->prox = tmp;
    
    if(primeiro == ultimo){
        ultimo = tmp;
    }
    tmp = NULL;
}


void removerInicio(){
    
    if(primeiro == ultimo){
        printf("erro ao remover!");
        return;
    }
    
    Celula* tmp = primeiro;
    primeiro = primeiro->prox;
    tmp->prox = NULL;
    free(tmp); // sempre quando inultilizar um ponteiro é preciso dar um free para liberar a alocação de memoria
    tmp = NULL;
    
}


void removerFim(){
    
      if(primeiro == ultimo){
        printf("erro ao remover!");
        return;
      }
    
    Celula* i;
    for( i = primeiro; i->prox != ultimo; i = i->prox);
    
    ultimo = i;
    free(ultimo->prox);
    i = ultimo->prox = NULL;
 
}

void mostrar(){
    
    printf("[");
    for(Celula* j = primeiro->prox; j != NULL; j = j->prox){
        
        printf("%d ", j->elemento);
    }
    printf("]");
    
}

int main()
{
    start(); //dar inicio a lista;
    
    inserirInicio(75);
    inserirInicio(83);
    inserirInicio(12);
    inserirInicio(54);
    inserirInicio(56);
    mostrar();
    
    
    
    


    return 0;
}

File: 2º Periodo/AEDS-II/Códigos/JAVA/ListaFlexivel.java
class Celula {
    public int elemento;
    public Celula prox;

    public Celula() {
        this(0);
    }

    public Celula(int elemento) {
        this.elemento = elemento;
        this.prox = null;
    }
}

class Lista {

    private Celula primeiro, ultimo;

    public Lista() {
        primeiro = new Celula();
        ultimo = primeiro;
    }
    
    
    
    public void inserirFim(){
        
        ultimo.prox = new Celula();
        ultimo = ultimo.prox;
        
    }
    
    public void inserirInicio(int x){
        
      Celula Tmp = new Celula(x);
      tmp.prox = primeiro.prox;
      primeiro.prox = tmp;
       
       if(primeiro == ultimo){
           ultimo = tmp;
       } 
       
       tmp = null;
        
    }
    
    public void inserir(int x, int pos){
        
      if(pos < 0 || pos > tamanho){
          throw new Exception("erro");
      }  
        
       if(pos == 0){
           inserirInicio(x);
       }else if(pos == tamanho){
           inserirFim(x);
       }
        
       Celula i = primeiro;
       for(int j = 0; j != pos; j++, i = i.prox);
       
       
        Celula tmp = new Celula(x);
        tmp.prox = i.prox;
        i.prox = tmp;
        tmp = null;       
     
    }
    

      public void removerInicio(){
       
      if(primeiro == ultimo){
          
          throw new Exception("erro, Lista vazia");
      }  
          
      Celula tmp = primeiro;
      primeiro = primeiro.prox;
      tmp.prox = null;
      tmp = null;
   
    }
    
    public void removerFim(){
        
     if(primeiro == ultimo){
         throw new Exception("erro lista vazia");
     } 
     
        
    Celula i = new Celula();    
        
    for(i = primeiro; i != ultimo; i = i.prox);
        
    int elemento = ultimo.elemento;   
        
    ultimo = tmp;
    tmp = tmp.prox = null;
    
    
    System.out.println(elemento);
        
    }
   
    public void remover(int pos){
        
        if(pos < 0 || pos > tamanho || primeiro == ultimo){
            throw new Exception("erro");
        }
        
        if(pos == primeiro){
            removerInicio();
        }else if(pos == tamanho){
            removerFim();
        }
        
        
        Celula i =primeiro;
        for(int j = 0; j != pos; j++ i = i.prox);
        
        Celula tmp = i.prox;
        i.prox = tmp.prox;
        tmp.prox = null;
        i = tmp = null 
        
        
        
        
        
        
        
    }


    public void mostrar{
        
        
     System.out.println("["); 
    
     for(Celula i = primeiro; i != ultimo; i = i.prox){
         
        System.out.printf(i.elemento + " "); 
         
     }
     
     System.out.println("]");    
      
    }
    
    public void MaiorElemento(){
        
     int maior = 0;
    
     for(Celula i = primeiro; i != null; i = i.prox){
         
        if(maior < i.elemento){
            
            maior = i.elemento;
            
        } 
    
     }
     
     System.out.printf("maior elemento é: " + maior); 
      
    }
    

}

public class ListaFlexivel {
    public static void main(String[] args) {
    
        try {
            Lista Lista = new Lista();
            Lista.inserir(4);
            Lista.inserir(5);
            Lista.inserir(2);
            Lista.inserir(6);
            Lista.inserir(8);
            Lista.mostrar();
          
            
  

        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
       
        
    }
    
}
